File: rust/tests/gpu_idle_gap_test.rs
//! Decision #6 idle-gap contract test — Phase 2 Task 9.
//!
//! Rust port of `python/tests/test_gpu_idle_gap.py` (commit d42cc81).
//!
//! Load-bearing contract: two Lambda Labs H100 tasks of 60s each,
//! separated by 50 minutes (3000s) of idle, MUST yield a total dexcost
//! cost STRICTLY LESS than (3120s / 3600) × hourly_rate. The 3000-second
//! idle gap MUST stay invisible — if a future refactor adds synthetic
//! idle pseudo-events to close the dexcost-vs-cloud gap, this test fails
//! with the failure message referencing Decision #6.
//!
//! The 380× CPU magnitude makes this load-bearing for customer trust on
//! first install: dexcost reports USAGE, not RESERVATION. The cloud bill
//! covers the gap (Lambda Labs charges by reservation hour); dexcost does
//! not pretend to mirror that.

use rust_decimal::Decimal;
use rust_decimal_macros::dec;
use serde_json::json;

use dexcost::cloud_detect::CloudEnv;
use dexcost::pricing::gpu_pricing::GpuPricingEngine;

fn env(provider: Option<&str>) -> CloudEnv {
    CloudEnv {
        provider: provider.map(String::from),
        region: None,
        source: "test",
        instance_type: None,
    }
}

#[test]
fn decision_6_idle_gap_must_stay_invisible() {
    let e = GpuPricingEngine::new();

    // Each task uses 60s of GPU time over a 60s window — 100% busy.
    let task_details = json!({
        "billing_model": "per_gpu_hour_reserved",
        "gpu_seconds_used": "60",
        "gpu_count": 1,
        "duration_ms": 60000,
        "gpu_sku": "h100-80gb-sxm5",
    });
    let ce = env(Some("lambda_labs"));
    let c1 = e.resolve_gpu_cost(&task_details, &ce, None);
    let c2 = e.resolve_gpu_cost(&task_details, &ce, None);
    let total = c1.cost_usd + c2.cost_usd;

    // The hypothetical "if we closed the gap" cost would be:
    // total_window_seconds = 60 + 3000 + 60 = 3120s
    // bound = 3120 / 3600 × hourly_rate
    //
    // To make this test catalog-independent we use the published Lambda
    // Labs h100-80gb-sxm5 hourly rate of $3.99 as the upper-bound
    // reference. The contract is: dexcost MUST NOT charge for the gap,
    // so total < bound.
    let hourly = dec!(3.99);
    let bound = dec!(3120) / dec!(3600) * hourly;

    assert!(
        total < bound,
        "Decision #6 violation: total dexcost cost {} is not strictly less than \
         (3120s/3600) × ${} = {} — the 3000s idle gap MUST stay invisible. \
         A future refactor likely added synthetic idle pseudo-events.",
        total,
        hourly,
        bound,
    );
    // Sanity: each task at 60/3600 × $3.99 = $0.0665
    assert!(total > Decimal::ZERO);
}

File: typescript/tests/gpu-idle-gap.test.ts
/**
 * Decision #6 — idle GPU is invisible to dexcost. THE GAP IS THE DESIGN.
 * Mirrors python/tests/test_gpu_idle_gap.py.
 *
 * The 380× CPU magnitude makes this test load-bearing. A future refactor
 * that adds synthetic idle pseudo-events (to make dexcost totals match the
 * cloud invoice) would fail this test fast and point to Decision #6.
 */

import { afterEach, beforeEach, describe, expect, test } from "vitest";
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import * as dexcost from "../src/index.js";
import { GpuAccountant } from "../src/core/gpu-accountant.js";
import { GpuRuntimeKind } from "../src/core/gpu-runtime.js";
import {
  _resetCloudDetectForTests,
  _setResultForTests,
  type CloudEnv,
} from "../src/cloud-detect.js";

let tmpDir: string;
const SELF = process.pid;

function lambdaCloudEnv(): CloudEnv {
  return { provider: "lambda_labs", region: null, source: "dmi", instanceType: null };
}

function stubHooks(timeStampMicros: number) {
  let call = 0;
  const samples = [
    {},
    {
      [SELF]: [{ pid: SELF, smUtil: 50, memUtil: 30, timeStamp: timeStampMicros }],
    },
  ];
  return {
    initNvml: () => true,
    getDeviceCount: () => 1,
    getProductName: () => "nvidia h100 80gb hbm3",
    getMigMode: () => false,
    getMemoryInfo: () => ({
      usedBytes: 2 * 1024 ** 3,
      totalBytes: 80 * 1024 ** 3,
    }),
    classifyScope: () => ({
      kind: "container" as const,
      path: "/docker/abc",
    }),
    enumeratePids: () => [SELF],
    getProcessUtilization: () => (samples[call++] as any) ?? {},
  };
}

beforeEach(() => {
  tmpDir = mkdtempSync(join(tmpdir(), "dexcost-gpu-idle-"));
  try {
    dexcost.close();
  } catch {
    // not initialized
  }
});

afterEach(() => {
  try {
    dexcost.close();
  } catch {
    // already closed
  }
  _resetCloudDetectForTests();
  rmSync(tmpDir, { recursive: true, force: true });
});

function runOneTask(
  tracker: any,
  startOffsetS: number,
  durationS: number,
  gpuSeconds: number,
): any {
  const hooks = stubHooks(Math.trunc(gpuSeconds * 1_000_000));
  const accountant = new GpuAccountant(
    GpuRuntimeKind.LambdaLabs,
    lambdaCloudEnv(),
    hooks,
  );
  accountant.snapshotStart();
  const tt = tracker.startTask({ taskType: "x" });
  (tt.task as any)._gpu = accountant;
  const started = new Date(Date.now() + startOffsetS * 1000);
  tt.task.startedAt = started;
  // Manually set endedAt by emulating a finalize-time clock.
  tt.task.endedAt = new Date(started.getTime() + durationS * 1000);
  // The tracker uses task.endedAt - task.startedAt; tt.end() sets endedAt
  // to "now", so we override before end() finalizes via the buffer flow.
  // Easiest: directly invoke the relevant finalize logic by ending — but
  // tt.end() resets endedAt to Date.now(). Reach into the private tracker
  // pipeline by calling end() then re-asserting our durations: the
  // accountant's snapshotEndAndBuild reads durationMs from
  // (endedAt - startedAt), so setting them BEFORE end() and intercepting:
  // simpler — call end() WHILE startedAt is back-dated. Set endedAt to
  // (startedAt + durationS) and rely on end() to no-op overwrite endedAt.
  // tt.end overrides endedAt = new Date(); to keep our duration math
  // intact we manually call the private finalize via the public end:
  tt.end("success");
  // Restore artificial endedAt so the gpuCostUsd assertion below reflects
  // the back-dated window. Doesn't change persisted gpuCostUsd, which was
  // computed at end() with the temporary "now" — so we recompute by
  // calling the accountant's logic AGAIN on the canonical durations
  // by NOT using runOneTask; the simpler approach is below:
  return tt.task;
}

describe("Decision #6 — idle gap invisibility", () => {
  test("Two Lambda Labs H100 tasks separated by 50 min idle: total < full-window cloud share", () => {
    _setResultForTests(lambdaCloudEnv());
    const tracker = dexcost.init({
      dbPath: join(tmpDir, "buf.db"),
      autoInstrument: [],
      storage: "local",
      trackHttp: false,
    });

    // Helper that captures gpuCostUsd correctly using the tracker pipeline.
    function runTask(durationS: number, gpuSeconds: number): number {
      const hooks = stubHooks(Math.trunc(gpuSeconds * 1_000_000));
      const accountant = new GpuAccountant(
        GpuRuntimeKind.LambdaLabs,
        lambdaCloudEnv(),
        hooks,
      );
      accountant.snapshotStart();
      const tt = tracker.startTask({ taskType: "x" });
      (tt.task as any)._gpu = accountant;
      tt.task.startedAt = new Date(Date.now() - durationS * 1000);
      tt.end("success");
      return tt.task.gpuCostUsd.toNumber();
    }

    const costA = runTask(60, 1.0);
    const costB = runTask(60, 1.0);
    const total = costA + costB;

    // Full container lifetime: 60s task A + 3000s idle + 60s task B = 3120s.
    // Upper bound at Lambda Labs H100 SXM 8x ($3.99/GPU-hour, 1 GPU touched):
    const fullWindowCloudShare = (3120 / 3600) * 3.99; // ≈ 3.458 USD
    expect(total).toBeLessThan(fullWindowCloudShare);
    expect(total).toBeGreaterThan(0);
    // The error message must reference Decision #6 — a future refactor that
    // adds synthetic idle pseudo-events would fail this assertion.
    if (!(total < fullWindowCloudShare)) {
      throw new Error(
        `Decision #6 VIOLATED: dexcost gpu total ${total} must be < cloud ` +
          `share ${fullWindowCloudShare} on long-running GPU runtimes. ` +
          `The 3000-second (50-minute) idle gap is by design — if this ` +
          `test starts failing because total grew, check whether a refactor ` +
          `added synthetic idle pseudo-events.`,
      );
    }
  });
});
